ABDM changes for consuming APIs in the mobile app.

akashkj opened the pull request:
## Product Description
Changes include:
1. Expose API to be consumed by ABDM android application. These API endpoints internally communicate with the ABDM server to return the processed response to the caller ABDM android app.
2. Authenticate new APIs with a custom token, which are having validity for ABDM API only.
3. Provide ABDM auth token to the CommCare mobile app user on every sync. There is the future scope of adding validity to the token.

## Technical Summary
https://docs.google.com/document/d/1gbQcAC3H4kj22iMCMMKSYMVO5_7kt56qiS8OWkd6e34/edit

## Feature Flag
No

## Safety Assurance

### Safety story
Manual testing in the local environment and dev instance of CommCare HQ deployed on Mumbai region.

### Automated test coverage
[TODO]Automated unit tests

### QA Plan
Pending


### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

mjriley committed:
remove safe from conditional alerts

mjriley committed:
Removed safe from SMS

mjriley committed:
Removed safe from accept_invite

mjriley committed:
Removed safe usage from users

mjriley committed:
Removed safe usages from HQAdmin

mjriley committed:
remove safe from domain templates

mjriley committed:
Remove safe filter from report templates

mjriley committed:
Removed safe filter from feature flags

mjriley committed:
Removed the safe filter from userreports templates

mjriley committed:
Removed safe filter from fixtures

mjriley committed:
Removed safe filter from accounting

mjriley committed:
Remove safe filter from custom data fields

mjriley committed:
remove safe filter from openmrs

mjriley committed:
Remove safe filter from inddex

mjriley committed:
remove careplan report templates

mjriley committed:
removed safe filter from casexml

mjriley committed:
Removed safe filter from app manager templates

mjriley committed:
Removed safe filter from crispy text field

mjriley committed:
removed safe filter from crispy templates

mjriley committed:
Removed unnecessary import

mjriley committed:
fixed a variable naming issue

mjriley committed:
Merge branch 'master' into mjr/safe_hqwebapp

mjriley committed:
Removed additional safe filters from templates

mjriley committed:
Updated invite template style

mjriley committed:
Merge branch 'mjr/safe_hqadmin' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/safe_app_manager' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/safe_custom_data_fields' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/safe_motech' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/safe_reports' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/safe_users' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/safe_sms' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/safe_casexml' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/safe_accounting' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/safe_fixtures' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/safe_toggleui' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/safe_domain' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/safe_messaging' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/safe_hqwebapp' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/safe_userreports' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/report_careplan' into mjr/mark_safe_combined

mjriley committed:
Merge branch 'mjr/safe_accept_invite' into mjr/mark_safe_combined

mjriley committed:
Updated DataTables output to be HTML-safe

mjriley committed:
Merge branch 'mjr/safe_inddex' into mjr/all_safe_fixes

mjriley committed:
Break out username for easier traceback handling in signup

mjriley committed:
Fixed an issue for new users attempting to accept an invite

mjriley committed:
Fixed conditional alert HTML artifact

mjriley committed:
Fixed an HTML artifact bug with custom reports

millerdev committed:
Merge branch 'master' into mjr/all_safe_fixes

Conflicts:
	corehq/apps/userreports/views.py

orangejenny committed:
Merge branch 'master' into mjr/all_safe_fixes

millerdev committed:
Postgresql 14

Previous version was 10.17

kaapstorm committed:
CommCare HQ

millerdev committed:
Log service version info in test output

Commit:
ABDM api for mobile app consumption

Commit:
Add abdm api token to response of mobile restore

Commit:
Add auth settings + register in primary urls.py

Commit:
abdm migrations

Commit:
add abdm exclusion to domain migration tests

snopoke commented on settings.py:
```diff
@@ -2015,6 +2020,9 @@ def _pkce_required(client_id):
 
 REST_FRAMEWORK = {
     'DATETIME_FORMAT': '%Y-%m-%dT%H:%M:%S.%fZ',
+    'DEFAULT_AUTHENTICATION_CLASSES': (
+       'rest_framework.authentication.TokenAuthentication',
+   ),
```
is this needed? it seems like you are specifying auth classes for all of your views already

akashkj replied:
Custom auth class is inheriting from `TokenAuthentication` and overrides `authenticate` method. This is then called by the framework as part of decorator `@authentication_classes((ABDMUserAuthentication,))` in the views.

snopoke replied:
Yes, I see that, but if you are specifying the auth class on all the views then there is no need to set a default for the whole of CommCare.

snopoke commented on custom/abdm/milestone_one/views/abha_verification_views.py:
```diff
@@ -0,0 +1,68 @@
+def get_auth_methods(request):
+    health_id = request.query_params.get("health_id")
+    if not health_id:
+        error_msg = "Missing required parameter in the request param: health_id"
```
could use `@required_request_params(["health_id"])`

akashkj replied:
This is a GET request and required_request_params is for POST request params. Didn't add extra decorator/logic as this is the only GET request.

snopoke commented on custom/abdm/models.py:
```diff
@@ -0,0 +1,32 @@
+    @property
+    def is_authenticated(self):
+        """
+        Always return True. This is a way to tell if the user has been
+        authenticated in templates.
```
```suggestion
        Always return True. This is a way to tell if the user has been
        authenticated in REST views.
```

snopoke commented on custom/abdm/milestone_one/utils/user_util.py:
```diff
@@ -0,0 +1,15 @@
+    user, _ = ABDMUser.objects.get_or_create(username=username)
+    if not user.access_token:
+        user.generate_token()
+    print(f"Received token {len(user.access_token)}")
```
nit: print

akashkj replied:
Removed

snopoke commented on custom/abdm/auth.py:
```diff
@@ -0,0 +1,19 @@
+from custom.abdm.models import ABDMUser
+
+
+class UserAuthentication(TokenAuthentication):
```
nit: rename to `ABDMUserAuthentication`

akashkj replied:
Done

snopoke commented on migrations.lock:
```diff
@@ -145,6 +147,10 @@ auth
  0010_alter_group_name_max_length
  0011_update_proxy_permissions
  0012_alter_user_first_name_max_length
+authtoken
+ 0001_initial
+ 0002_auto_20160226_1747
+ 0003_tokenproxy
```
Are these needed? I don't see any usages of DRF AuthTokens

akashkj replied:
Custom auth class is inheriting from `TokenAuthentication` and overrides `authenticate` method. This is then called by the framework as part of decorator `@authentication_classes((ABDMUserAuthentication,))` in the views.

snopoke commented on settings.py:
```diff
@@ -234,6 +234,7 @@
     'django_user_agents',
     'logentry_admin',
     'oauth2_provider',
+    'rest_framework.authtoken',
```
Is this used?

akashkj replied:
Custom auth class is inheriting from `TokenAuthentication` and overrides `authenticate` method. This is then called by the framework as part of decorator `@authentication_classes((ABDMUserAuthentication,))` in the views.

snopoke replied:
`TokenAuthentication` is part of the base `rest_framework` library so if that's all you are using then you don't need this.

snopoke reviewed (commented):
One general comment: After a [lengthy discussion](https://github.com/dimagi/commcare-hq/issues/31160) on the use of Type Hints in commcare code the prevailing opinion is not to use them.

akashkj commented:
> One general comment: After a [lengthy discussion](https://github.com/dimagi/commcare-hq/issues/31160) on the use of Type Hints in commcare code the prevailing opinion is not to use them.

Ok, will remove it. Wasn't aware of this. Thanks.

akashkj commented on corehq/ex-submodules/casexml/apps/phone/xml.py:
```diff
@@ -129,10 +130,16 @@ def get_registration_element_data(restore_user):
         "uuid": restore_user.user_id,
         "date": date_to_xml_string(restore_user.date_joined),
-        "user_data": restore_user.user_session_data
+        "user_data": get_session_data_with_abdm_token(restore_user)
```
Is it possible to check for a particular application, rather than a domain? For example, if some property is set within an application, then only the key will be returned. 

Will check the extensions part.

snopoke replied:
> Is it possible to check for a particular application, rather than a domain? For example, if some property is set within an application, then only the key will be returned.

That could be possible. You would need to change the `get_registration_element` to accept more arguments (potentially just pass in the whole `RestoreState` class)

Commit:
Merge remote-tracking branch 'origin/master' into akj/abdm-api-for-mobile-app

akashkj commented on custom/abdm/milestone_one/utils/decorators.py:
```diff
@@ -0,0 +1,32 @@
+    def decorate(fn):
+        @wraps(fn)
+        def wrapped(request, *args, **kwargs):
+            if not (params and isinstance(params, List)):
+                error_msg = f"Request {request} could not be validated as validation input not provided."
+                logger.warning(error_msg)
+                return generate_invalid_req_response(error_msg)
```
Done

akashkj commented on custom/abdm/milestone_one/utils/decorators.py:
```diff
@@ -0,0 +1,32 @@
+                return generate_invalid_req_response(error_msg)
+            invalid_params = []
+            for param in params:
+                if not request.data.get(param):
```
Done

akashkj commented on custom/abdm/models.py:
```diff
@@ -0,0 +1,32 @@
+
+class ABDMUser(models.Model):
+    id = models.AutoField(primary_key=True)
+    username = models.CharField(max_length=100)
```
Yes, that will be the case. Modified.

akashkj commented on custom/abdm/milestone_one/utils/request_util.py:
```diff
@@ -0,0 +1,29 @@
+    if additional_headers:
+        headers.update(additional_headers)
+    resp = requests.post(url=req_url, data=json.dumps(payload), headers=headers)
+    return resp.json()
```
Added

Commit:
Refactor based on review comments

joxl committed:
Change the way a "for export" adapter is acquired

gherceg committed:
remove CRUDPaginatedViewMixin from AutomaticUpdateRuleListView

- add new JS and HTML files that handle pagination and crud operations
- use tabs to prepare for adding a new history tab
- decouple DeduplicationRuleListView from AutomaticUpdateRuleListView for now

snopoke commented on custom/abdm/auth.py:
```diff
@@ -0,0 +1,19 @@
+        if not token:
+            return None
+        try:
+            user = ABDMUser.objects.get(access_token=token)
```
If you make this model comply with the rest framework key structure then you could just do this:
```
class ABDMUserAuthentication(TokenAuthentication):
    model = ABDMUser
```

A rest framework token has a 'key' and 'user' fields (user is a FK to the Django user).

I'm also unclear why you need a custom model for the token and can't either use the existing HQApiKey or else the DRF Token model.

akashkj replied:
Intention is to restrict token's access to ABDM API only as it is accessed by a separate app. Will it possible to implement this limitation without a custom auth logic?

akashkj replied:
I remember that if those entries related to DRF's token were not added in the `settings.py` file, it resulted in the authentication system not getting triggered.

snopoke replied:
ok

gherceg committed:
fix stickler JS warnings

joxl committed:
Change document adapter usage pattern

- Adapters are now module-level instances (rather than classes to be
  initialized when needed)
- Adapters are registered when created (for future verification at app
  setup time)

joxl committed:
Move `index_name` and `type` class attributes to init args

gherceg committed:
first batch of addressing feedback

joxl committed:
Update the manager adapter pattern to match document adapters

Use a single manager instance rather than instantiating everywhere it's
needed.

gherceg committed:
second batch of addressing feedback

gherceg committed:
ignore stickler syntax error

gherceg committed:
unwrap id in delete rule modal

gherceg committed:
move header and help text outside of tabs

joxl committed:
Overhaul `ElasticDocumentAdapter.update()`

gherceg committed:
use page_context, not main_context

gherceg committed:
basic implementation of rule run history tab

gherceg committed:
rename list_automatic_update_rules.js to deduplication_rules.js

joxl committed:
Refactor `ElasticDocumentAdapter` keyword arguments

Remove uneeded `**kw` and leave notes for others that should be removed

gherceg committed:
duplicate format_rule code into deduplication list view (the irony)

planning to do refactoring the deduplication view in the near
future which will remove duplicate code

gherceg committed:
use get_timezone_for_user directly

gherceg committed:
add paginated for rule run history tab

gherceg committed:
missing time from page context

gherceg committed:
add num of updates and closes to table

gherceg committed:
Merge branch 'master' into gh/replace-base-crud-view

gherceg committed:
filter DomainCaseRuleRun by case update workflow

gherceg committed:
add status field to table

gherceg committed:
refactor to ensure 'update rule' is not used

AmitPhulera committed:
misc: remove older django version compatibility

AmitPhulera committed:
misc: more docstring in transient_util

AmitPhulera committed:
add: Tombstone class

AmitPhulera committed:
tests: Tomobstone class

AmitPhulera committed:
refactor:set default  property to None

AmitPhulera committed:
refactor: some refactors to support the multiplexer class coming up

AmitPhulera committed:
add: ElasticMultiplexAdapter class

AmitPhulera committed:
refactor: create_document_adapter to return multiplexer class if secondary index is provided

AmitPhulera committed:
refactor: split helper functions into a sepearate test helper class

AmitPhulera committed:
add: tests for the multiplexer adapter

gherceg committed:
add allowed_actions to dedupe view

gherceg committed:
Merge branch 'gh/replace-base-crud-view' into gh/add-history-tab

AmitPhulera committed:
add docstrings for tombstones

AmitPhulera committed:
update readme as per the changes in the PR

gherceg committed:
rename reference to auto_update_rule_run_history.html

AmitPhulera committed:
move index cleanup from teardown to self.addCleanup

AmitPhulera committed:
Update corehq/apps/es/transient_util.py

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

AmitPhulera committed:
pass fn instead of result in addCleanup

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

gherceg committed:
sort rule runs by descending started and finished dates

nospame committed:
Add checkbox case search option to app manager

AmitPhulera committed:
document usage for tombstones in readme

nospame committed:
Add checkbox to case search query view

nospame committed:
Add checkbox prompt to xml tests

nospame committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into em/case-search-checkbox

AmitPhulera committed:
add mgmnt command to get es version when index was created

AmitPhulera committed:
remove confusing para from readme

nospame committed:
Refactor

Simplify changeQueryField logic, remove changeDateQueryField as duplicative, use tuple for list of appearance types

gherceg committed:
bump default items per page from 5 to 25

SmittieC committed:
Shadow module to get persistent and inline details from itself instead of its source module

SmittieC committed:
Refactor a bit

minhaminha committed:
Nitpicking

AddisonDunn committed:
remove unused methods

AddisonDunn committed:
add TableauUser, TableauGroup models

Commit:
Merge remote-tracking branch 'origin/master' into akj/abdm-api-for-mobile-app

AddisonDunn committed:
add translations for tableau roles

nospame committed:
Revert "Refactor"

This reverts commit 94436e6d3264b5a055247391f9036768b69dd7c5.

SmittieC committed:
refactor - look for self.app on the EntriesHelper instead of passing it in as arguement

gherceg committed:
changed default pagination to 25 on wrong table

nospame committed:
Use lookup tables in checkbox case search property

nospame committed:
Allow multiple checkboxes in case search UI

nospame committed:
Stickler

Semicolons, use fat arrow for single line return

snopoke committed:
initial command to backfill subevent date

snopoke committed:
update from sms

snopoke committed:
backfill from xformsession

snopoke committed:
make chunksize configurable

snopoke committed:
add migration status to command

snopoke committed:
test that rows with dates are not updated

snopoke committed:
combine tests into a single class

AddisonDunn committed:
increase max length for username field

Commit:
Merge remote-tracking branch 'origin/master' into akj/abdm-api-for-mobile-app

kaapstorm committed:
Merge pull request #32151 from dimagi/nh/commcare_hq_rst

"CommCareHQ" -> "CommCare HQ"

AmitPhulera committed:
Merge pull request #32289 from dimagi/ap/es-index-version-created

Add management command to get es version when index was created

snopoke committed:
add more options for testing

snopoke committed:
fix backfill tests

snopoke committed:
Revert "[SAAS-14102] addf field to messaging sub-event model to record date of last activity"

snopoke committed:
Merge pull request #32314 from dimagi/revert-32307-sk/subevent-date-last-activity

Revert "[SAAS-14102] addf field to messaging sub-event model to record date of last activity"

snopoke committed:
Revert "Revert "[SAAS-14102] addf field to messaging sub-event model to record date of last activity""

snopoke committed:
set 'auto_now' in a separate migration to avoid table lock on the initial migration

snopoke committed:
improve query performance

snopoke committed:
add count query for better progress & refactor

AddisonDunn committed:
Merge pull request #32302 from dimagi/ad/create-other-tableau-syncing-models

Create other tableau syncing models

snopoke committed:
skip if no rows

snopoke committed:
lower defualt chunksize

snopoke committed:
Merge pull request #32315 from dimagi/revert-32314-revert-32307-sk/subevent-date-last-activity

Revert "Revert "[SAAS-14102] addf field to messaging sub-event model to record date of last activity""

snopoke committed:
Merge branch 'master' into sk/backfill-subevent-date

millerdev committed:
Optimize EnterpriseFormReport.total_for_domain()

Use count() query rather than len(<hits>) because the latter was timing out for some domains.

Commit:
Feature flag for restore content

Commit:
Use extensions for fetching abdm token in restore

SmittieC committed:
Merge pull request #32298 from dimagi/cs/sc-2369/shadow_module_fix

Fix: Shadow module to get its persist context and inline details from its own module during build

snopoke committed:
Merge pull request #32311 from dimagi/sk/backfill-subevent-date

[SAAS-14102]  backfill subevent "date_last_activity"

snopoke committed:
update API to use new date field for sorting

millerdev committed:
Merge pull request #32317 from dimagi/dm/optimize-enterprise-forms-report

Optimize EnterpriseFormReport.total_for_domain()

millerdev committed:
Merge branch 'master' into dm/pg14

millerdev committed:
DRY postgres version and warn on outdated pin

millerdev committed:
Merge pull request #32120 from dimagi/dm/pg14

PostgreSQL 14

minhaminha committed:
Update staging.yml

orangejenny committed:
Updated EOF nav config to use select2

minhaminha committed:
Merge pull request #32288 from dimagi/mjr/all_safe_fixes

Remove unnecessary safe filters within templates

nospame committed:
Fix html typo

nospame committed:
Show multiselect disabled + checked for checkbox property

nospame committed:
updating files: scripts/staging.yml

[ci skip]

gherceg committed:
Merge pull request #32211 from dimagi/gh/replace-base-crud-view

Remove paginated crud view mixin from auto update rule view

gherceg committed:
Update staging.yml

dannyroberts committed:
Update translations

[ci skip]

AmitPhulera committed:
Merge pull request #32265 from dimagi/jm+ap/elastic-multiplexer

ElasticSearch Multiplexer Adapter.

orangejenny committed:
Merge pull request #32323 from dimagi/jls/eof-nav-select2s

Updated EOF nav config to use select2

gherceg committed:
only check for data registry if not remote link

gherceg committed:
add search box to rule run table

gherceg committed:
add brief description of rule run history table

nospame committed:
Merge pull request #32290 from dimagi/em/case-search-checkbox

Add checkbox to case search fields

gherceg committed:
use datasource var in data registry check

the datasource var is populated whether the link is local or remote
and can be used to verify if this is a data registry report

gherceg committed:
datasource as optional param in is_data_registry_report

gherceg committed:
Update staging.yml

gherceg committed:
update description for rule run history table

Co-authored-by: Jenny Schweers <orange.jenny@gmail.com>

joxl committed:
Merge pull request #32324 from dimagi/create-pull-request/update-translations-d291ae3

Update Translations

esoergel committed:
updating files: scripts/staging.yml

[ci skip]

gherceg committed:
Merge pull request #32326 from dimagi/gh/resolve-remote-link-issue

Only check for data registry if not remote domain link

gherceg committed:
Update staging.yml

gherceg committed:
Merge pull request #32258 from dimagi/gh/add-history-tab

Add auto update rule history tab

gherceg committed:
Update staging.yml

snopoke committed:
update staging

snopoke committed:
Merge pull request #32322 from dimagi/sk/messaging-api-new-date-field

[USH-2606] update API to use new date field for sorting

Commit:
Merge branch 'akj/abdm-api-for-mobile-app' of https://github.com/dimagi/commcare-hq into akj/abdm-api-for-mobile-app

Commit:
remove drf default auth from settings.py

akashkj commented:
Opened https://github.com/dimagi/commcare-hq/pull/32332
